### PR TITLE
Issue 28265 cli support sending workflow names when saving content types

### LINKED ISSUE #28265: CLI: Support sending workflow names when saving content types
_Transcript from Will's requirements:_ 

When specifying workflows, the user should be able to pass just an array of workflow names OR ids e.g.
```
  "workflows" : [ 
     "System Workflow", 
     "d61a59e1-a49c-46f2-a929-db2b4bfa88b2"
  ]
```
See the content Type Def below that I used to test:
```
{
  "clazz" : "SimpleContentType",
  "name" : "Testing CLI Type",
  "variable" : "testingCLIType",
  "folder" : "SYSTEM_FOLDER",
  "siteName" : "demo.dotcms.com",
  "folderPath" : "/",
  "icon" : "blender",
  "baseType" : "CONTENT",
  "fields" : [ {
    "clazz" : "TextField",
    "indexed" : true,
    "listed" : true,
    "name" : "Title",
    "variable" : "title",
    "fieldType" : "Text",
    "contentTypeId" : "c676e84ad83b82c12e3d264a2200408d",
    "dataType" : "TEXT"
  }, {
    "clazz" : "TagField",
    "indexed" : true,
    "name" : "Keywords",
    "variable" : "keywords",
    "fieldType" : "Tag",
    "contentTypeId" : "c676e84ad83b82c12e3d264a2200408d",
    "dataType" : "SYSTEM"
  }, {
    "clazz" : "TextField",
    "searchable" : true,
    "indexed" : true,
    "listed" : true,
    "name" : "Url Slug",
    "variable" : "urlSlug",
    "fieldType" : "Text",
    "contentTypeId" : "c676e84ad83b82c12e3d264a2200408d",
    "dataType" : "TEXT"
  }, {
    "clazz" : "BinaryField",
    "name" : "Image",
    "variable" : "image",
    "fieldType" : "Binary",
    "contentTypeId" : "c676e84ad83b82c12e3d264a2200408d",
    "dataType" : "SYSTEM"
  }, {
    "clazz" : "StoryBlockField",
    "name" : "Blog",
    "variable" : "blog",
    "fieldType" : "Story-Block",
    "contentTypeId" : "c676e84ad83b82c12e3d264a2200408d",
    "dataType" : "LONG_TEXT"
  } ],
  "detailPage" : "f0ffaa29bf4a5ae46f127fd4d6acb55f",
  "urlMapPattern" : "/ai-blog/{urlSlug}/",
  "workflows" : [ {
    "archived" : false,
    "creationDate" : "2024-03-05T16:18:34.738+00:00",
    "defaultScheme" : false,
    "description" : "",
    "entryActionId" : null,
    "id" : "d61a59e1-a49c-46f2-a929-db2b4bfa88b2",
    "mandatory" : false,
    "modDate" : "2024-03-05T12:02:58.553+00:00",
    "name" : "System Workflow",
    "system" : true
  } ]
}

```

### PR DESCRIPTION
**Pull request by jgambarios**
Now, when updating the content type workflows, we don't depend solely on the workflow ID, but we can use now the workflow variable name as a secondary key if needed.

To accomplish this, we needed to introduce in the workflow scheme the concept of the variable name (Like the variable we have in content types and fields, same concept).

This PR fixes: #28265

### CODE REVIEW

File: dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task240606AddVariableColumnToWorkflow.java [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+import com.dotmarketing.exception.DotRuntimeException;
+import com.dotmarketing.startup.StartupTask;
+import com.dotmarketing.util.Logger;
+import com.dotmarketing.util.StringUtils;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+public class Task240606AddVariableColumnToWorkflow implements StartupTask {
```
Comment (fabrizzio-dotCMS) [inline dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task240606AddVariableColumnToWorkflow.java:14]: Shouldnt this have a test

File: dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java [hunk 2]
```diff
@@ -606,13 +608,13 @@ public WorkflowScheme findSystemWorkflowScheme() throws DotDataException {
 
 	@Override
 	@CloseDBIfOpened
-	public WorkflowScheme findScheme(final String id) throws DotDataException, DotSecurityException {
+	public WorkflowScheme findScheme(final String idOrVar) throws DotDataException, DotSecurityException {
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java:611]: shouldn't we have a findScheme test by variable in the WorkflowAPITest?

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/contenttype/ContentTypeForm.java [hunk 3]
```diff
@@ -102,68 +112,140 @@ private List<List<Tuple2<WorkflowAPI.SystemAction, String>>> systemActionWorkflo
...
+                        WORKFLOW_ATTRIBUTE_NAME);
+
+                for (Object entry : workflowsJSONArray) {
+
+                    final WorkflowFormEntry workflowFormEntry;
+
+                    if (entry instanceof String) { // Entry is a workflow id
+
+                        workflowFormEntry = WorkflowFormEntry.builder().
+                                id((String) entry).
```
[the remark this answers is not in the document]
Comment (jgambarios) [reply dotCMS/src/main/java/com/dotcms/rest/api/v1/contenttype/ContentTypeForm.java:206]: Done

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/contenttype/ContentTypeForm.java [hunk 4]
```diff
@@ -183,17 +265,63 @@ private static List<Tuple2<WorkflowAPI.SystemAction, String>> getSystemActionsWo
         }
     }
 
+    /**
+     * Represents a workflow entry found in the form
+     */
+    public static class WorkflowFormEntry {
```
[the remark this answers is not in the document]
Comment (jgambarios) [reply dotCMS/src/main/java/com/dotcms/rest/api/v1/contenttype/ContentTypeForm.java]: Done

File: dotCMS/src/main/java/com/dotmarketing/business/DeterministicIdentifierAPIImpl.java [hunk 5]
```diff
@@ -560,6 +569,19 @@ private boolean isCategoryId(final String hash){
                 .getInt("test")>0;
     }
 
+    /**
+     * Test the calculated hash has already been used as a workflow identifier
+     *
+     * @param hash the hash to test
+     * @return true if the hash has been used as a workflow identifier
+     */
+    private boolean isWorkflowId(final String hash) {
```
[the remark this answers is not in the document]
Comment (jgambarios) [reply dotCMS/src/main/java/com/dotmarketing/business/DeterministicIdentifierAPIImpl.java:587]: Done

File: dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task240606AddVariableColumnToWorkflow.java (hunk 1 above)
Comment (jgambarios) [reply dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task240606AddVariableColumnToWorkflow.java:14]: Done

File: dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java (hunk 2 above)
Comment (jgambarios) [reply dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java:611]: Done

**Review by fabrizzio-dotCMS — APPROVED** [review]

File: dotcms-integration/src/test/java/com/dotmarketing/startup/runonce/Task240606AddVariableColumnToWorkflowTest.java [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+/**
+ * Test class for {@link Task240606AddVariableColumnToWorkflow}.
+ * <p>
+ * This test class contains test cases to validate the addition of the variable column to the
+ * workflow_scheme table and to handle scenarios where workflows with duplicate names exist.
+ */
+public class Task240606AddVariableColumnToWorkflowTest {
```
Comment (nollymar) [inline dotcms-integration/src/test/java/com/dotmarketing/startup/runonce/Task240606AddVariableColumnToWorkflowTest.java:27]: A reference of this test class is missing in any of the MainTestSuites. It won't be executed.
Besides, the test javadocs should follow this convention https://auth.dotcms.com/devwiki/integration-unit-testing-conventions

File: dotcms-integration/src/test/java/com/dotmarketing/portlets/workflows/business/WorkflowAPITest.java [hunk 7]
```diff
@@ -4663,4 +4674,128 @@ public void test_comment() throws DotDataException, DotSecurityException {
         assertEquals("comment test 1", comments1.get(0).commentDescription());
     }
 
+    /**
```
Comment (nollymar) [inline dotcms-integration/src/test/java/com/dotmarketing/portlets/workflows/business/WorkflowAPITest.java:4677]: The test javadocs should follow this convention https://auth.dotcms.com/devwiki/integration-unit-testing-conventions

**Review by nollymar — APPROVED** [review]

File: dotcms-integration/src/test/java/com/dotmarketing/portlets/workflows/business/WorkflowAPITest.java (hunk 7 above)
Comment (jgambarios) [reply dotcms-integration/src/test/java/com/dotmarketing/portlets/workflows/business/WorkflowAPITest.java:4677]: Done

File: dotcms-integration/src/test/java/com/dotmarketing/startup/runonce/Task240606AddVariableColumnToWorkflowTest.java (hunk 6 above)
Comment (jgambarios) [reply dotcms-integration/src/test/java/com/dotmarketing/startup/runonce/Task240606AddVariableColumnToWorkflowTest.java:27]: Done

**Pull request merged**